icu4c/source/common/ulocimp.h

```diff
@@ -307,4 +307,75 @@ U_CAPI const char* const* ulocimp_getKnownCanonicalizedLocaleForTest(int32_t* le
 // Return true if the value is already canonicalized.
 U_CAPI bool ulocimp_isCanonicalizedLocaleForTest(const char* localeName);
 
+/**
+ * A utility class for handling locale IDs that may be longer than ULOC_FULLNAME_CAPACITY.
+ * This encompasses all of the logic to allocate a temporary locale ID buffer on the stack,
+ * and then, if it's not big enough, reallocate it on the heap and try again.
+ *
+ * You use it like this:
+ * UErrorCode err = U_ZERO_ERROR;
+ *
+ * PreflightingLocaleIDBuffer tempBuffer;
+ * do {
+ *     *(tempBuffer.getRequestedCapacityPtr()) = uloc_doSomething(localeID, tempBuffer.getBuffer(), tempBuffer.getCapacity(), &err);
```

Comment:
why not just a void requestCapacity(int32_t newCapacity) ?

Reply:
Well, the original idea was to keep it looking as much like how you'd normally use one of these functions as possible:

```
newCapacity = uloc_doSomething(...);
```

My concern with using a setter...

```
tempBuffer.requestCapacity(uloc_doSomething(...));
```

...was that it obscured the important part of the line, the `uloc_doSomething()` call, making it harder to see what the calling function was doing and putting the emphasis on all the crap we're having to do to make sure we have an appropriately-sized character buffer (the complaint against my original fix for this problem).

That said, I'm not sure `*(tempBuffer.getRequestedCapacityPtr) =` is really doing that.  It's weird-looking and long, and both of those work against the goal of emphasizing the `uloc_doSomething()` part of the line.  `tempBuffer.requestedCapacity = ...` would be better, but I didn't like the idea of making an internal member public here.

The more I think about it, the more I think making `requestedCapacity` public and getting rid of that weird getter would be better, combined with some extra validity checking in `needToTryAgain()`.  What do you guys think?

Reply:
> making `requestedCapacity` public and getting rid of that weird getter would be better, ... What do you guys think?

FWIW, I'd be inclined to agree. I think that it might help to simplify things.

For example, instead of this:
`*(tempBuffer.getRequestedCapacityPtr()) = uloc_doSomething(localeID, tempBuffer.getBuffer(), tempBuffer.getCapacity(), &err);`
it might look something like this:
`tempBuffer.requestedCapacity = uloc_doSomething(localeID, tempBuffer.getBuffer(), tempBuffer.getCapacity(), &err);`
 
Since it's an internal helper class, making it public seems fine.

Reply:
Okay, I went ahead and made that change.  Thanks a lot!

If this looks good to you, I'll squash on your say-so while we're waiting for Frank.

Reply:
Ah, I hate to kept making nit comments on your PR.... but the example in the comment block in `ulocimp.h` still has the method `getRequestedCapacityPtr()`  :)

```c++
*(tempBuffer.getRequestedCapacityPtr()) = uloc_doSomething(localeID, tempBuffer.getBuffer(),
```

Reply:
Ugh.  Sorry about that.  Fixed now.